### PR TITLE
Usage Stats: Add metrics to count enabled kms providers per kind

### PR DESCRIPTION
**Pull request by joanlopez**
**What this PR does / why we need it**:

It adds metrics to count current providers per kind and enabled kms providers per kind.

**Which issue(s) this PR fixes**:

Contributes to https://github.com/grafana/grafana-enterprise/issues/2377

**Special notes for your reviewer**:

Note that the strategy followed consists on encapsulating the logic to extract the provider kind from provider identifier into a specific type `secrets.ProviderID`. 

It implied updating the name of the default encryption provider from `secretKey` to `secretKey.v1` to keep it as a consistent & valid provider identifier. That could be considered kinda breaking change but should be fine because the feature is under a feature toggle.

Alternatively, I also thought about a slightly different strategy with a `NewProviderID` constructor (which could be a bit safer / less error prone) and make the `Kind()` method errorless, but I was not completely sure that's an accepted convention in our code base (cc/ @sakjur).

Thanks! 🙌🏻

### CODE REVIEW

File: pkg/services/secrets/secrets.go [hunk 1]
```diff
@@ -42,3 +44,16 @@ type Provider interface {
...
 }
+
+type ProviderID string
+
+func (id ProviderID) Kind() (string, error) {
+	idStr := string(id)
+
+	parts := strings.SplitN(idStr, ".", 2)
+	if len(parts) != 2 {
+		return "", fmt.Errorf("malformatted provider identifier %s: expected format <provider>.<keyName>", idStr)
```
Comment (sakjur) [inline pkg/services/secrets/secrets.go:55]: Doesn’t this break backwards compatibility?

File: conf/defaults.ini [hunk 2]
```diff
@@ -235,7 +235,7 @@ admin_password = admin
 secret_key = SW2YcwTIb9zpOOhoPsMm
 
 # current key provider used for envelope encryption, default to static value specified by secret_key
-encryption_provider = secretKey
+encryption_provider = secretKey.v1
```
Comment (sakjur) [inline conf/defaults.ini:238]: …because the old value for this field should still remain valid, if it has been present in a release.

**Review by sakjur — COMMENTED** [review]
Unsure about the backwards compatibility of this change.

Also remove usage insights from the title, this is unrelated to usage insights (cannot remember right now what the metric collection service is called, but usage insights is for user facing features).

**Comment by sakjur** [comment]
Jumped straight into a review 😅

> It implied updating the name of the default encryption provider from secretKey to secretKey.v1 to keep it as a consistent & valid provider identifier. That could be considered kinda breaking change but should be fine because the feature is under a feature toggle.

The ini changes are not under a feature toggle, so I’m not sure it’s ok.

> Alternatively, I also thought about a slightly different strategy with a NewProviderID constructor (which could be a bit safer / less error prone) and make the Kind() method errorless, but I was not completely sure that's an accepted convention in our code base (cc/ @sakjur).

:+1: I think that’s fine, we’re not very dogmatic about conventions as long as they’re not eyesores or unnecessarily complex. Personally I’m fond of bespoke types with custom validation logic and would like to see it for things like IDs everywhere to reduce the risk of flipping i.e. orgID and userID.

**Comment by sakjur** [comment]
Can we handle the “secret_key” value specifically as equivalent to “secret_key.v1”?

**Comment by joanlopez** [comment]
> Unsure about the backwards compatibility of this change.

Yup, unsure as well. Thought about taking the opportunity that (apparently) no one is using this feature and it's still behind a feature toggle, to tidy it up (using a consistent & valid identifier; extending the possibility of using different secret keys -to migrate them, for instance-, etc).

Otherwise I can see how and where to add an edge case for the default identifier.
Maybe processing `secretKey` as `secretKey.v1` when reading settings could be enough, what do you think? 🤔 

> Also remove usage insights from the title, this is unrelated to usage insights (cannot remember right now what the metric collection service is called, but usage insights is for user facing features).

Ouch, I'm super bad at naming things... after some minor explorations I think it's "Usage Stats" instead of "Usage Insights".
Updated, thanks for letting me know 😊 Correct me again if still wrong, please 🙏🏻 

Finally, what do you think about the latest comment on the description (where I tagged you) 🤔 I'm curious about your feedback on that 😛 🙈

**Comment by joanlopez** [comment]
> Can we handle the “secret_key” value specifically as equivalent to “secret_key.v1”?

Just seen we thought the same concurrently, going for that! 🚀  Thanks! 🙌🏻

**Comment by joanlopez** [comment]
> The ini changes are not under a feature toggle, so I’m not sure it’s ok.

Yeah, fair, for some reason I thought it was commented out of the box (which could still be undesired kind of breaking change). But definitely not :(

> 👍 I think that’s fine, we’re not very dogmatic about conventions as long as they’re not eyesores or unnecessarily complex. Personally I’m fond of bespoke types with custom validation logic and would like to see it for things like IDs everywhere to reduce the risk of flipping i.e. orgID and userID.

Yeah, the usual reflection is:

- Okay, let's use custom types (`OrgID`, `UserID`, etc) instead of basic types (`int`, `string`, etc).
- But.. then no one prevents you from doing (`OrgID(...)` or directly "123" when `OrgID`) is backend by a string, which makes that type a bit useless (we could consider it does not reduces how error prone a code is due to that reason).
- Then.. let's make it safer, let's use a struct based type with private fields, a `New` constructor and getter methods, but... (a) that does not sound super idiomatic; (b) sounds a bit of overengineering and (c) no one prevents you from `OrgID{}` (although in this case it's much easier to detect in a quick code review).
- So.. let's go back to basic types?

That's why I usually do prefer using basic types rather than falling in the middle of this reflection. However, in certain situations, it completely makes sense and it's pretty helpful to use a custom type to encapsulate some logic and / or add some checks / validations. Then it's when you fall into that flow of reflections and there's when team conventions could help, I guess.

**Comment by sakjur** [comment]
That’s a really insightful loop of “why would anyone do X when you can do Y when you can do Z when you can do X”. Reminds me of https://bellmar.medium.com/hunting-tech-debt-via-org-charts-92df0b253145 that I stumbled on from Stack Overflow’s newsletter the other week.

I think the first step (use a type that requires intent) is sufficient for an improvement. The second step is where people err. The idea is that it’s the responsibility of the code to prevent programmers from introducing bugs— I don’t think it is. It’s the responsibility of the code to nudge programmers slightly, to require intent. That’s why we have types, errors, and things like it.

Slippery slopes are hard, because we cannot program them away, we have to have developers who reason about these things like you’re doing now. And maybe improve our internal comms re: any such principle we adopt across the codebase.

File: pkg/services/secrets/secrets.go (hunk 1 above)
Comment (joanlopez) [reply pkg/services/secrets/secrets.go:55]: In theory, [these changes](https://github.com/grafana/grafana/pull/43640/commits/ee0ed8fba31ed6886b2ff0de11c3faa52c6fb674) should fix it. WDYT? 🤔

**Review by sakjur — APPROVED** [review]

**Comment by joanlopez** [comment]
cc/ @mjseaman @undef1nd  for visibility. Added metrics are:

- `stats.encryption.envelope_encryption_enabled.count` --> to count how many Grafana instances have envelope encryption enabled
- `stats.encryption.current_provider.PROVIDER_KIND.count` (per each provider kind -awskms, azurekv, etc-) --> to count how Grafana many instances use each provider kind as the "current" (if so, or if they rely on the default one 🙈)
- `stats.encryption.envelope_encryption_enabled.count` --> to count how many Grafana instances have envelope encryption enabled
- `stats.encryption.providers.PROVIDER_KIND.count` (per each provider kind -awskms, azurekv, etc-) --> to count the average of providers enabled per kind per Grafana instance.

**Pull request merged**